Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -184,36 +185,80 @@ attribute [to_additive UniqueSums] UniqueProds
+  and addition is strictly monotone w.r.t. the second argument, then `A` has `UniqueSums`." ]
+instance (priority := 100) Covariant.to_uniqueProds_right {A} [Mul A] [IsRightCancelMul A]
+    [LinearOrder A] [CovariantClass A A (· * ·) (· < ·)] :
+    UniqueProds A where
+  uniqueMul_of_nonempty {A B} hA hB := by
+    obtain ⟨a0, b0, ha0, hb0, he⟩ := Finset.mem_mul.mp (Finset.max'_mem _ <| hA.mul hB)
+    refine ⟨a0, ha0, b0, hb0, fun a b ha hb he' => ?_⟩
+    obtain hl | rfl | hl := lt_trichotomy b b0
+    · refine ((he'.trans he ▸ mul_lt_mul_left' hl a).not_le <| Finset.le_max' _ (a * b0) ?_).elim
+      exact Finset.mem_mul.mpr ⟨a, b0, ha, hb0, rfl⟩
+    · exact ⟨mul_right_cancel he', rfl⟩
+    · refine ((he ▸ mul_lt_mul_left' hl a0).not_le <| Finset.le_max' _ (a0 * b) ?_).elim
+      exact Finset.mem_mul.mpr ⟨a0, b, ha0, hb, rfl⟩
+
+-- see Note [lower instance priority]
+/-- This instance asserts that if `A` has a left-cancellative multiplication, a linear order,
+  and multiplication is strictly monotone w.r.t. the first argument, then `A` has `UniqueProds`. -/
+@[to_additive Covariant.to_uniqueSums_left
+  "This instance asserts that if `A` has a left-cancellative addition, a linear order,
+  and addition is strictly monotone w.r.t. the first argument, then `A` has `UniqueSums`." ]
+instance (priority := 100) Covariant.to_uniqueProds_left {A} [Mul A] [IsLeftCancelMul A]
+    [LinearOrder A] [CovariantClass A A (Function.swap (· * ·)) (· < ·)] :
+    UniqueProds A where
+  uniqueMul_of_nonempty {A B} hA hB := by
+    obtain ⟨a0, b0, ha0, hb0, he⟩ := Finset.mem_mul.mp (Finset.max'_mem _ <| hA.mul hB)
+    refine ⟨a0, ha0, b0, hb0, fun a b ha hb he' => ?_⟩
+    obtain hl | rfl | hl := lt_trichotomy a a0
+    · refine ((he'.trans he ▸ mul_lt_mul_right' hl b).not_le <| Finset.le_max' _ (a0 * b) ?_).elim
+      exact Finset.mem_mul.mpr ⟨a0, b, ha0, hb, rfl⟩
+    · exact ⟨rfl, mul_left_cancel he'⟩
+    · refine ((he ▸ mul_lt_mul_right' hl b0).not_le <| Finset.le_max' _ (a * b0) ?_).elim
+      exact Finset.mem_mul.mpr ⟨a, b0, ha, hb0, rfl⟩
+
+namespace UniqueProds
+variable {G H : Type*} [Mul G] [Mul H]
+
+@[to_additive (attr := nontriviality, simp)]
+theorem of_subsingleton [Subsingleton G] : UniqueProds G :=
+  ⟨fun ⟨a, ha⟩ ⟨b, hb⟩ => ⟨a, ha, b, hb, UniqueMul.of_subsingleton⟩⟩
+
+open Finset in
+@[to_additive]
+theorem mulHom_image_of_injective (f : H →ₙ* G) (hf : Function.Injective f) (uG : UniqueProds G) :
+    UniqueProds H := by
+  refine ⟨fun {A B} A0 B0 => ?_⟩
+  classical
+  obtain ⟨a0, ha0, b0, hb0, h⟩ := uG.uniqueMul_of_nonempty (A0.image f) (B0.image f)
+  rcases mem_image.mp ha0 with ⟨a', ha', rfl⟩
+  rcases mem_image.mp hb0 with ⟨b', hb', rfl⟩
+  exact ⟨a', ha', b', hb', (UniqueMul.mulHom_image_iff f hf).mp h⟩
+
+/-- `UniqueProd` is preserved under multiplicative equivalences. -/
+@[to_additive "`UniqueSums` is preserved under additive equivalences."]
+theorem mulHom_image_iff (f : G ≃* H) :
```

Comment:
Honestly, I do not really know if this iff is useful or not: I thought that I needed it for the step "identify a ℚ-vector space with the `Finsupp basis → ℚ`", but the injection version was enough.  This is a fairly common pattern: equivalences in formal maths are less ubiquitous than in informal maths.

Reply:
Yeah but I wouldn't bother removing it ...

Reply:
Agreed.  It could also be useful for rewrites.